Bare minimum arg completion for scala3

tanishiking opened the pull request:
Related https://github.com/scalameta/metals-feature-requests/issues/112

Though there're still some things to do, I'm submitting a draft PR because I want to ask for advice around the failing tests :pray:, or want to at least have this as a reference implementation for people who want to implement arg completion in the future.

![scala3-arg-minimum](https://user-images.githubusercontent.com/9353584/103643155-8c73fe00-4f97-11eb-9cd0-57ddda7e7662.gif)

I think it's ideal to fix all tests implemented in `cross`, but it might be ok to ship it because it's better than nothing 😅

---

## Some test cases that are failing
### arg / arg9
```scala
object Main {
  1.until(@@)
}
```

We get empty completion because `methodSym` in `ArgCompletion` always `NoSymbol`.
I guess this is because `assert(@@)` and `1.until(@@)` have multiple overloaded methods like `until(end: Long)` and `until(start: Long, end: Long)`, and that's why we can't retrieve a specific symbol.

```scala
val method = apply.fun
val methodSym = method.symbol
```

It looks like scala2 implementation somehow deals with this case (as `arg9`), but I have no idea how to deal with it in scala3 🤔 
(How to retrieve at least one possible symbol for the method).

### arg4
```scala
case class User(
    name: String = "John",
    age: Int = 42,
    address: String = "",
    followers: Int = 0
)
object Main {
  User("", @@, address = "")
}
```

- in scala2 implementation, the completion suggests `age` and `followers`
- in this impl, `age`, `address`, and `followers`
  - because `args` in ArgCompletion is `List(Literal(Constant("")), Select(Ident(User),$lessinit$greater$default$3), Select(Ident(User),$lessinit$greater$default$4))`, and the nemed arg `address` looks disappeared 🤔 :

### arg6
```scala
case class User(
    name: String = "John",
    age: Int = 42,
    address: String = "",
    followers: Int = 0
)
object Main {
  User("", @@ "")
}
```

- in scala2 implementation, the completion suggests `age`, `followers`, and `followers`
- in this impl `address` and `followers`
  - because `args` in ArgCompletion is `List(Literal(Constant("")), Literal(Constant("")), Select(Ident(User),$lessinit$greater$default$3), Select(Ident(User),$lessinit$greater$default$4))` and the second argument is still there, that's why `name` and `age` are dealed with as filled args.

## Things we're missing that implemented for scala2 (not to do in this PR)
- filtering and sorting completions
- `findPossibleDefaults` from scope
- `fillAllFields`

tgodzik commented on mtags/src/main/scala-3/scala/meta/internal/pc/Completions.scala:
```diff
@@ -0,0 +1,104 @@
+    .zip(baseParams.iterator)
+    // filter out synthesized default args
+    .filterNot {
+      case (Ident(name), _) => name.toString.contains("$")
```
I wonder if we should indeed do this? Maybe Scala 3 works differently here? This will also cause people to not be able to use named params that contain `$`, which I think is a bug and is reported under https://github.com/scalameta/metals/issues/2348

tanishiking replied:
Yeah, in e745df8 tried no to use `$` for filtering out synthetic params and arg getters.

- we could filter out synthesized param like `evidence$1` for something like `def foo[T: Ordering](...)`.
  - https://github.com/scalameta/metals/pull/2369/commits/e745df83da8341f420fd8b176d226ca4509393b9#diff-020d74d58479b87b103a4b80f6ac22e6dd510390fa221baad3305cba2b321649R75
- and we could filter out default args by using `DefaultGetterName` instead of just checking name contains `$` or not.
  - https://github.com/lampepfl/dotty/blob/a5d3ba96b4fd9bcefe3d6d1e7ef69453fe1c0fee/compiler/src/dotty/tools/dotc/core/NameKinds.scala#L328

Maybe we could filter out synthesized params by using `Symbol#isStynthetic` in scala2 :)

tgodzik commented on mtags/src/main/scala-3/scala/meta/internal/pc/ScalaPresentationCompiler.scala:
```diff
@@ -124,7 +124,10 @@ case class ScalaPresentationCompiler(
-          Completion.completions(pos)(using ctx.fresh.setCompilationUnit(unit))._2
+          val path = Interactive.pathTo(driver.openedTrees(uri), pos)(using ctx)
+          val completions = Completion.completions(pos)(using ctx.fresh.setCompilationUnit(unit))._2
+          val completions2 = completionPosition(pos, path)(using ctx)
```
```suggestion
          val metalsCompletions = completionPosition(pos, path)(using ctx)
```
?

tgodzik commented on tests/cross/src/test/scala/tests/pc/CompletionArgSuite.scala:
```diff
@@ -114,7 +143,14 @@ class CompletionArgSuite extends BaseCompletionSuite {
+    topLines = Option(3),
+    compat = Map(
+      "3.0" ->
+        """|address = : String
```
It looks the same as the default, so maybe compat is not needed? I think I noticed it in a couple of other places

tanishiking replied:
Ah, that's correct, fixed in  https://github.com/scalameta/metals/pull/2369/commits/8d2a018299384c70dff7e64dc64edcf16d2b3a52

tgodzik reviewed (changes requested):
Looks really good! I left some comments and as for the failing tests:

> arg9

This might be good to report in Dotty, I don't think anything  should return `NoSymbol`. This might be something that was fixed in Scala 2 a while ago.

> arg4

I think this might also be problem to fix in Dotty itself. Or at least discussed with the compiler team.

> arg6

I wonder if it might be connected to the fact that in Scala 2 we add the artificial `CURSOR` to the interpreted source code. I don't really remember why it was needed, but it might be the reason.

tgodzik commented:
Anyway, I am fine ignoring the tests that are not working currently (it is still improvement over it not working currently :sweat_smile: ) - we should just follow up with issues in Metals and maybe Dotty in some cases.

tanishiking commented:
Thank you so much for the review!
Added some commits for fixing your comments, and skipped failing tests in https://github.com/scalameta/metals/pull/2369/commits/4331986b0ba6acb8490f725dd1a09094161578a2

Now it's ready to review, and I'm also fine with shipping this leaving falling tests (because they require much more effort to fix, working on dotty compiler maybe).

tgodzik commented on tests/cross/src/test/scala/tests/pc/CompletionArgSuite.scala:
```diff
@@ -44,7 +51,12 @@ class CompletionArgSuite extends BaseCompletionSuite {
        |""".stripMargin,
-    topLines = Option(3)
+    topLines = Option(3),
+    compat = Map( // findPossibleDefaults and fillAllFields not yet supported
```
This is not an issue in this test case I think so maybe remove the comment? Same above and below.

tgodzik commented on tests/cross/src/test/scala/tests/pc/CompletionArgSuite.scala:
```diff
@@ -114,7 +145,8 @@ class CompletionArgSuite extends BaseCompletionSuite {
        |""".stripMargin,
-    topLines = Option(3)
+    topLines = Option(3),
+    assumeCond = Some(!isScala3)
```
Let's add a comment here, why it's not working.

tgodzik commented on tests/cross/src/test/scala/tests/pc/CompletionArgSuite.scala:
```diff
@@ -227,7 +280,8 @@ class CompletionArgSuite extends BaseCompletionSuite {
        |argument = argument : Int
        |""".stripMargin,
-    topLines = Some(3)
+    topLines = Some(3),
```
Should it not be just:
```scala
"""|argument: Int
    |argument = : Int
""".stripMargin
```

tgodzik commented on tests/cross/src/test/scala/tests/pc/CompletionArgSuite.scala:
```diff
@@ -247,7 +301,13 @@ class CompletionArgSuite extends BaseCompletionSuite {
+        """|argument = : Int
+           |""".stripMargin
+    ),
+    assumeCond = Some(!isScala3)
```
Why ignore it? Shouldn't it work currently?

tgodzik commented on tests/cross/src/test/scala/tests/pc/CompletionArgSuite.scala:
```diff
@@ -261,7 +321,8 @@ class CompletionArgSuite extends BaseCompletionSuite {
     "foo(rele@@)",
-    "foo(relevant)"
+    "foo(relevant)",
+    assumeCond = Some(!isScala3)
```
This should work currently normally.

tanishiking replied:
Unfortunately, no it doesn't work in scala3.
Because `relevant` is searched from the scope, but scala3 doesn't support it yet.

tgodzik reviewed (changes requested):
Left some comments, thanks for the great work! We could refine it more later. Scala 3 is still not stable, so it's fine to be a bit experimental.

tanishiking commented:
@tgodzik Thanks for the review!
I will delve into the dotty internal and file an issue later :)

tgodzik commented:
@tanishiking is it ready to rereview?

tanishiking commented:
@tgodzik Oops, sorry yes  it's ready for review :)

tgodzik commented on tests/cross/src/main/scala/tests/BasePCSuite.scala:
```diff
@@ -33,10 +34,18 @@ abstract class BasePCSuite extends BaseSuite {
   val executorService: ScheduledExecutorService =
     Executors.newSingleThreadScheduledExecutor()
   val scalaVersion = BuildInfoVersions.scalaVersion
+  val isScala3: Boolean = isScala3Version(scalaVersion)
```
This is no longer used.

tgodzik commented on tests/cross/src/main/scala/tests/BasePCSuite.scala:
```diff
@@ -33,10 +34,18 @@ abstract class BasePCSuite extends BaseSuite {
   protected val workspace = new TestingWorkspaceSearch
   val tmp: AbsolutePath = AbsolutePath(Files.createTempDirectory("metals"))
 
+  implicit class RichTestOptions(options: TestOptions) {
```
And I think this also?

tgodzik commented on tests/cross/src/test/scala/tests/pc/CompletionArgSuite.scala:
```diff
@@ -31,7 +31,12 @@ class CompletionArgSuite extends BaseCompletionSuite {
-    topLines = Option(3)
+    topLines = Option(3),
+    compat = Map(
+      "3.0" ->
```
This can be removed since we ignore the test.

tanishiking replied:
Actually, this `compat` is for test `arg1` which is not ignored.
(I'm also confused because GitHub folded the code  😅 )

tgodzik replied:
Ach! Sorry about that!

tgodzik commented on tests/cross/src/main/scala/tests/BasePCSuite.scala:
```diff
@@ -238,6 +247,9 @@ abstract class BasePCSuite extends BaseSuite {
     }
   }
 
+  object IgnoreScala3Version
```
```suggestion
  object IgnoreScala3
```
since it ignores it all?

tgodzik commented on tests/cross/src/test/scala/tests/pc/CompletionArgSuite.scala:
```diff
@@ -99,11 +118,17 @@ class CompletionArgSuite extends BaseCompletionSuite {
        |""".stripMargin,
-    topLines = Option(4)
+    topLines = Option(4),
+    compat = Map(
```
Compat section can be removed here also.

tanishiking replied:
It's for `arg5` not `arg4`

tgodzik replied:
My brain filtered out the line numbers :sweat:

tgodzik commented on tests/cross/src/test/scala/tests/pc/CompletionArgSuite.scala:
```diff
@@ -212,7 +252,12 @@ class CompletionArgSuite extends BaseCompletionSuite {
        |isResourceFile = isLargeBanana : Boolean
-       |""".stripMargin
+       |""".stripMargin,
+    compat = Map(
```
If the test case is ignored we can remove the compat section.

tanishiking replied:
This is for `arg14` not for `arg9`

tgodzik reviewed (changes requested):
Looks good! We should be able to merge it, but I would remove some of a leftover code.

Thanks again for the great contribution!

tanishiking commented:
@tgodzik Thank you so much for reviewing!
Removed leftover code as your feedback (sry, I left a lot of unnecessary code 😅 ), renamed `s/IgnoreScala3Version/IgnoreScala3/`.

I left some comments around `compat`, please take a look :pray:

tgodzik reviewed (approved):
Thanks for removing the leftover code and sorry about those wrong comments. I totally missed that!

tgodzik commented:
`An error occurred while performing auto-merge.` - this is great :sweat_smile:
